File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        ans = []
        left = 0
        right = len(matrix[0])
        top = 0
        bot = len(matrix)
        while left <= right and bot >= top:
            for j in range(left,right):
                ans.append(matrix[top][j])
            top += 1
            #last column
            for i in range(top,bot):
                ans.append(matrix[i][right-1])
            right -= 1
            #last row
            for j in range(right-1, left-1, -1):
                ans.append(matrix[bot-1][j])
            bot -= 1
            #fisrt column
            for i in range(bot-1, top-1, -1):
                ans.append(matrix[i][left])
            left += 1
        return ans[:len(matrix[0])*len(matrix)]

File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> ans;
        int left = 0, right = matrix[0].size() - 1;
        int up = 0, down = matrix.size() - 1;
        while(left<=right && up <=down ){
            
            for(int i = left; i<=right; i++){
                ans.push_back(matrix[up][i]);
            }
            up++;
            for(int j = up; j<=down; j++){
                ans.push_back(matrix[j][right]);
            }
            right--;
            if(up <= down){
                for(int i = right; i>=left; i--){
                    ans.push_back(matrix[down][i]);
                }
                down--;
            }

            if(left <= right){
                for(int j = down; j>=up; j--){
                    ans.push_back(matrix[j][left]);
                }
                left++;
            }
            
            
        }
        return ans;
    }
};